File: LC0139_Word_Break/word_break.py
from typing import List

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        wordDict_set = set(wordDict)
        n = len(s)
        # dp[i] = if we can segment s[i...n - 1] into a combination of wordDict's words
        dp = [False] * (n + 1)
        dp[n] = True
        max_word_length = max(len(word) for word in wordDict)

        for i in range(n - 1, -1, -1):
            for j in range(i, min(i + max_word_length, n)):
                # dp[j + 1] represents if we can properly segment everything in s after index j
                # If so, then we check if the substring from i to j (both inclusive) can be properly segmented
                if dp[j + 1] and s[i: j + 1] in wordDict_set:
                    dp[i] = True
                    break # Exit j loop

        return dp[0]

File: LC0139_Word_Break/word_break.go
package main

func wordBreak(s string, wordDict []string) bool {
	wordDict_set := make(map[string]bool)
	n := len(s)
	// dp[i] = if we can properly segment the substring s[i...n - 1]
	dp := make([]bool, n+1)
	dp[n] = true
	max_word_length := -1
	for _, word := range wordDict {
		wordDict_set[word] = true
		x := len(word)
		if x > max_word_length {
			max_word_length = x
		}
	}
	for i := n - 1; i >= 0; i-- {
		for j := i; j < i+max_word_length && j < n; j++ {
			_, exists := wordDict_set[s[i:j+1]]
			if dp[j+1] && exists {
				dp[i] = true
				break // Exit j loop
			}
		}
	}

	return dp[0]
}
